sympy/matrices/numpy.py

```diff
@@ -0,0 +1,308 @@
+from __future__ import division, print_function
+
+from collections import Counter
+
+from sympy import Float
+from sympy.core.compatibility import default_sort_key
+from sympy.core.sympify import sympify
+from sympy.simplify import nsimplify
+from sympy.utilities.decorator import doctest_depends_on
+
+from .common import NonSquareMatrixError
+from .dense import MutableDenseMatrix
+
+_TOLERANCE   = 1e-13 # tolerance for nsympify rationalization
+_TOLERANCEIM = 1e-6 # much lower tolerance for truncating imaginaries to reals, set to 1e-5 if numerical errors persist
+
+
+def _truncim(x): # truncate very small imaginary component to real
+    return x if abs(x.imag) >= _TOLERANCEIM else x.real
+
+def _fsimplify(x):
+    return nsimplify(_truncim(x), rational=True, tolerance=_TOLERANCE)
+
+
+def _detect_dtype(l): # Is there a better way to do this?
+    import numpy as np
+
+    try:
+        f = [float(e) for e in l]
+    except TypeError: # complex or symbolic source
+        pass
+
+    else:
+        try:
+            i = [int(e) for e in l]
+        except TypeError: # have infinities and/or nans
+            return np.float64
+
+        else:
+            if any(isinstance(e, Float) for e in l) or i != f:
+                return np.float64
+            else:
+                return np.int64
+
+    try:
+        all(complex(e) for e in l)
+    except TypeError: # non-numeric source data
+        raise TypeError('source data for NumPyMatrix must be numeric') from None

```

Comment:
Can we do the same with `if-else`?

Reply:
Please elaborate... (Keep in mind that at this point the objects in the list could be anything, Python, SymPy, mpf objects, anything really it comes from the user).